service/src/main/java/bio/terra/workspace/service/resource/controlled/cloud/azure/AzureStorageAccessService.java

```diff
@@ -60,6 +64,10 @@ public class AzureStorageAccessService {
   private final AzureConfiguration azureConfiguration;
   private final WorkspaceService workspaceService;
   private final Map<StorageAccountCoordinates, StorageData> storageAccountCache;
+  private final Map<String, SamUser> samUserCache;
```

Comment:
Yeah, this is a lot of caches. But, looking through the code, this all feels like essential complexity:
* We need the sam user info to emit the audit log entry. This is general WSM tech debt I think (moving away from authenticateduserrequest)
* We need the storage container resource, which has an implicit sam check against the controlled resource, so we can get the container resource name for building the sas URL
* We need the permissions on the container to ensure the actions the user has available and that we're sanely building the sas token permissions

That middle one strikes me as awkward but I can't think of a better approach.